Password strength checker using JavaScript Regex (#1105)
section-io/engineering-education: articles/password-checker-javascript/index.md

|[a-z]|Checks for lowercase letters|
|[A-Z]|Checks for uppercase letters|
|x(?=y)|Returns x if and only if it is followed by y|
|**.**|Checks for any single character except line terminators |
|x*|Checks for x 0 or more times|

#### The JavaScript program

We are going to discuss how to check the strength of a password that a user enters based on:

- Is at least 8 characters long(`?=.{8,}`)

- Has at least one uppercase letter(`?=.*[A-Z]`)

- Has at least one lowercase letter(`?=.*[a-z]`)

- Has at least one digit(`?=.*[0-9]`)

- Has at least one special character(`[^A-Za-z0-9]`)

We are going to have three levels:

- **Strong**: The password has to meet all the requirements to be strong.

```java
(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])(?=.{8,})
```

![Strong password](/engineering-education/password-checker-javascript/strong-password.png)

- **Medium**: The password can only be medium either if it is at least 6 characters long and meets all the other requirements or has no digit but meets the rest of the requirements.

```java
((?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])(?=.{6,}))|((?=.*[a-z])(?=.*[A-Z])(?=.*[^A-Za-z0-9])(?=.{8,}))

> ```

Reviewer: @Agusioma Can you provide step by step construction of these Regular Expressions.

<html>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <!-- Bootstrap 5 link -->
    <link href= "https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">

<!-- CSS Styling -->
<style>
    .passwordInput{
        margin-top: 5%; 
        text-align :center;
    }

    .displayBadge{
        margin-top: 5%; 
        display: none; 
        text-align :center;
    }
</style>
    
<body>
    <p>
        <div class="d-flex text-danger justify-content-center">
            <ul style="list-style-type:none">
                <li class="fs-4" style="margin-top: 10%; text-align :center">
                    Password Strength Checker
                </li>
                <li>
                    <input id="0101" class="form-control passwordInput" type="password" placeholder="Type your password" id="inputPassword">
                </li>
                <li>
                    <span id="0102" class="badge displayBadge">Weak</span>
                </li>
            </ul>
        </div>  
    </p>     
</body>
</html>

> ```

Reviewer: Can you briefly explain what the HTML code does? The elements you are adding and why you are adding them.

#### JavaScript Code

You can take a look at the complete code in this [GitHub repository](https://github.com/Agusioma/Password-Strength-Checker/).

> ```java

Reviewer: In this section, can you go step by step on how to write the script and then provide the entire script at the end.

        strengthBadge.style.backgroundColor = 'blue';
        strengthBadge.textContent = 'Medium';
    } else {
        strengthBadge.style.backgroundColor = 'red';
        strengthBadge.textContent = 'Weak';
    }
}
```

Let's add an input event listener to the `password` input and check the strength with the `StrengthChecker` function when the user has typed something. We won't call the function immediately after every keystroke. When the user types quickly, we should wait until a pause occurs. So, instead of checking the strength immediately, we'll set a timeout.

We should clear the previous timeout if there is any. When the events occur close together than our timeout duration, the timeout from the preceding input event should be canceled.

```JavaScript
password.addEventListener("input", () => {
    strengthBadge.style.display = 'block';
    clearTimeout(timeout);
    timeout = setTimeout(() => StrengthChecker(password.value), 500);
    if(password.value.length !== 0) {
        strengthBadge.style.display != 'block';
    } else {
        strengthBadge.style.display = 'none';
    }
});
```

The full JavaScript code:

```JavaScript
<script>

>     

Reviewer: I'd suggest you create a new file for the JS code instead of writing it in a Script tag to keep things clean. 

You should instruct the users to create a new file to write the code and then link it using a script tag.

#### HTML Code and explanation

```HTML
<html>
<head>
    <!-- Bootstrap 5 link -->
    <link href= "https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">
    <!-- CSS Styling -->
    <style>
        .passwordInput{
            margin-top: 5%; 
            text-align :center;
        }

        .displayBadge{
            margin-top: 5%; 
            display: none; 
            text-align :center;
        }
    </style>
</head>
<body>
    <div class="d-flex text-danger justify-content-center">
        <h4 style="margin-top: 10%; text-align :center">
            Password Strength Checker
        </h4>
        <br>
        <input type="password" placeholder="Type your password" id="PassEntry" class="form-control passwordInput">
        <br>
        <span id="StrengthDisp" class="badge displayBadge">Weak</span>
        <br>
    </div>     
</body>
</html>
```

- We'll use Bootstrap 5 for styling. We'll use a CDN link for the bootstrap file and link it to the HTML file using the `<link>` tag. 

> - We'll write our custom styles using the `<style>` tag. We have two classes and their respective stylings.

Reviewer: Hi @Agusioma 
Minor suggestion. Add info about which elements you are writing the custom styles for.

    ```

![Strong password](/engineering-education/password-checker-javascript/strong-password.png)

- **Medium**:  If the password is at least 6 characters long and meets all the other requirements, or has no digit but meets the rest of the requirements.

    The code is the same as for the **Strong** level only that **?=.{6,}** shows that we are checking for at least 6 characters. It also has **|** for checking either of the two conditions `(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])(?=.{6,})` or `(?=.*[a-z])(?=.*[A-Z])(?=.*[^A-Za-z0-9])(?=.{8,})`.

    ```JavaScript
    ((?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])(?=.{6,}))|((?=.*[a-z])(?=.*[A-Z])(?=.*[^A-Za-z0-9])(?=.{8,}))
    ```

![Medium password](/engineering-education/password-checker-javascript/medium-password.png)

- **Weak**: If the entered password does not meet the strong or medium-level requirements, then it is deemed weak.

![Weak password](/engineering-education/password-checker-javascript/weak-password.png)

#### HTML Code and explanation

A badge is used to show the password strength and color. Green for strong, Blue for medium and Red for weak.

```HTML
<html>
    <!-- Bootstrap 5 link -->

>     <link href= "https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">

Reviewer: Wrap the `<link>` tag in the `<head>` tag.

<!-- CSS Styling -->
<style>
    .passwordInput{
        margin-top: 5%; 
        text-align :center;
    }

    .displayBadge{
        margin-top: 5%; 
        display: none; 
        text-align :center;
    }
</style>
    
<body>
    <div class="d-flex text-danger justify-content-center">
        <ul style="list-style-type:none">
            <li class="fs-4" style="margin-top: 10%; text-align :center">
                Password Strength Checker
            </li>
            <li>
                <input id="0101" class="form-control passwordInput" type="password" placeholder="Type your password" id="inputPassword">
            </li>
            <li>

>                 <span id="0102" class="badge displayBadge">Weak</span>

Reviewer: Use meaningful IDs instead of numbers. The input tag has 2 IDs, use only one.

            </li>
        </ul>

>     </div>     

Reviewer: There's no reason that all the elements must be in an `<ul>`. Keep it simple.

Just add everything inside the div without the `<ul>`.

```HTML
<div class="d-flex text-danger justify-content-center">
    <h4 class="fs-4" style="margin-top: 10%; text-align :center">Password Strength Checker</h4>
    <input id="0101" class="form-control passwordInput" type="password" placeholder="Type your password" id="inputPassword">
    <span id="0102" class="badge displayBadge">Weak</span>
</div>     
```

</body>
</html>
```

- We will use Bootstrap 5 for styling. We will provide a CDN link for the bootstrap file and link it to the HTML file using the `<link>` tag. 

- We write our custom styles using the `<style>` tag. We have two classes and their respective stylings.

- In the `<body>` tag we have a `<div>` element.

- In the `<div>`, we have an unordered list with the style `list-style-type:none` for removing bullets. It has three items (`<li>`):

- The first is under class `fs-4` which makes the text size that of **h4**. We have also given it an inline styling.

- The second item is an input where our password will be typed in. It is under the Bootstrap's class `form-control` and the class `passwordInput` which we custom styled in the `<style>` tag. The `form-control` shows the inputshould be used in a `form` element(We have not added a form since we are not sending any data).

- The third item contains a `<span>` which is where the password strength will be labeled. It is under the class `displayBadge` which we custom styled and `badge` given by Bootstrap which makes it appear as a badge.

#### JavaScript Code creation and explanation

> Create five variables: `timeout` for storing the timeout before a callback is called, `password` & `strengthBadge` for storing the input and span after traversing the DOM and getting them using their IDs, and `strongPassword` & `mediumPassword` for storing the Regex conditions.

Reviewer: ```suggestion
Let's create five variables: `timeout` for storing the timeout before the callback function is called, `password` & `strengthBadge` for storing the DOM element of the input and the span, and `strongPassword` & `mediumPassword` for storing the Regex conditions.
```

Reviewer: Say what the span is used for that you are selecting.

```JavaScript
let timeout;
let password = document.getElementById('0101')
let strengthBadge = document.getElementById('0102')
let strongPassword = new RegExp('(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])(?=.{8,})')
let mediumPassword = new RegExp('((?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])(?=.{6,}))|((?=.*[a-z])(?=.*[A-Z])(?=.*[^A-Za-z0-9])(?=.{8,}))')
```

> We then add an input event listener to the `password` input where we check the strength when the user has typed something, but we won't do it immediately after every input event. When typing quickly, we wait until a pause occurs so instead of immediately acting as the event handler, we set a timeout.

Reviewer: ```suggestion
Let's add an input event listener to the `password` input and check the strength when the user has typed something, but we won't do it immediately after every keystroke. When the user types quickly, we should wait until a pause occurs. So, instead of checking the strength immediately, we'll set a timeout.
```

> We also clear the previous timeout, if there is any so that when the events occur close together than our timeout duration, the timeout from the preceding input event is canceled.

Reviewer: ```suggestion
We should clear the previous timeout if there is any. When the events occur close together than our timeout duration, the timeout from the preceding input event should be canceled.
```

```JavaScript
password.addEventListener("input", () => {
    strengthBadge.style.display= 'block'
    clearTimeout(timeout)
    timeout = setTimeout(() => StrengthChecker(password.value), 500)
    if(password.value.length !== 0){
        strengthBadge.style.display != 'block'
    }else{
        strengthBadge.style.display = 'none'
    }
});
```

> We then create a function `StrengthChecker` where we test for the match through the `RegExp.prototype.test()` method. It returns true if there is a match or false if there is no match.

Reviewer: ```suggestion
Let's create a function called `StrengthChecker` where we test for the strength of the passwords using the regular expressions that we wrote with the `RegExp.prototype.test()` method. It returns `true` if there is a match or `false` if there is no match.
```

> It then sets the background colour (`strengthBadge.style.backgroundColor`) and text of the badge (`strengthBadge.textContent`). The `if else` statement is used in deciding what will be done from the results gotten.

Reviewer: ```suggestion
Then, let's set the background colour (`strengthBadge.style.backgroundColor`) and text of the badge (`strengthBadge.textContent`) according to the result. 
```
